### PR TITLE
78 return book frontend

### PR DESCRIPTION
**Pull request by harryBedfordSL**
**78 - Return Book Frontend**
- Added the front end for returning a book
- User is now warned with a "Are you sure" popup message to confirm returning of book
- Confirm button added with inverted colours
- Added reducer tests to make sure correct action is being invoked for each new action creator

### CODE REVIEW

File: src/main/client/src/dashBoard/bookList/book/Book.js [hunk 1]
```diff
@@ -32,21 +32,41 @@ export default class Book extends Component {
...
+            ? "green"
+            : "blue"
+          : null;
+        buttonText = book.availabilityChecked
+          ? book.isAvailable
+            ? "Borrow"
+            : "Reserve"
+          : "Check Availability";
+        break;
+    }
```
Comment (EllieHield) [inline src/main/client/src/dashBoard/bookList/book/Book.js]: probably ought to refactor my original bit to match your format for clarity :)

File: src/main/client/src/dashBoard/bookList/reducer.js [hunk 2]
```diff
@@ -127,6 +150,80 @@ export const checkBook = bookId => dispatch => {
...
+    if (user.borrows) {
+      user.borrows.forEach(borrow => {
+        if (borrow.bookId === book.id && borrow.active) {
+          borrowId = borrow.id;
+          reservationId = null;
+          role = "Borrower";
+          popupText = "Return your Book";
+        }
+      });
+    }
```
Comment (EllieHield) [inline src/main/client/src/dashBoard/bookList/reducer.js]: not sure the if statement is necessary? As long as user.borrows is initialised as an empty array
same for user.reservations

File: src/main/client/src/dashBoard/bookList/reducer.js [hunk 3]
```diff
@@ -77,35 +77,58 @@ export const deleteBook = bookIds => (dispatch, getState) => {
...
       dispatch(popupText("Reservation successful!", bookId));
+      const newBooks = getState().bookList.books.map(eachBook => {
+        if (eachBook.id === book.id) {
+          eachBook.borrowId = null;
+          eachBook.reservationId = res.data;
+          eachBook.role = "Reserver";
+          eachBook.popupText = "Cancel your Reservation";
+        }
+        return eachBook;
+      });
```
Comment (Ro4052) [inline src/main/client/src/dashBoard/bookList/reducer.js]: Maybe this should be a find

File: src/main/client/src/dashBoard/bookList/reducer.js [hunk 4]
```diff
@@ -77,35 +77,58 @@ export const deleteBook = bookIds => (dispatch, getState) => {
...
+      const newBooks = getState().bookList.books.map(eachBook => {
+        if (eachBook.id === book.id) {
+          eachBook.borrowId = res.data;
+          eachBook.reservationId = null;
+          eachBook.role = "Borrower";
+          eachBook.popupText = "Return your Book";
+          eachBook.returnStarted = false;
+        }
+        return eachBook;
+      });
```
Comment (Ro4052) [inline src/main/client/src/dashBoard/bookList/reducer.js]: Maybe these should use some sort of callback

File: src/main/client/src/dashBoard/bookList/book/requestButton/RequestButton.js [hunk 5]
```diff
@@ -2,22 +2,47 @@ import React, { Component } from "react";
...
+  constructor(props) {
+    super(props);
+    this.checkClick = this.checkClick.bind(this);
+  }
+
+  checkClick(evt) {
+    if (evt.target !== this.confirmButton.childNodes[0]) {
+      this.props.cancelProcess(this.props.book);
+    }
+    document.removeEventListener("click", this.checkClick);
```
Comment (Ro4052) [inline src/main/client/src/dashBoard/bookList/book/requestButton/RequestButton.js]: Does the listener get removed if you actually click confirm?

**Pull request merged**